CylindricalBinaryCompactObject: add ExcisionSpheres.

markscheel opened the pull request:
Adds ExcisionSpheres to CylindricalBinaryCompactObject.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
CylBinaryCompactObject: add ExcisionSpheres.

geoffrey4444 commented on src/Domain/Creators/CylindricalBinaryCompactObject.cpp:
```diff
@@ -801,7 +810,62 @@ Domain<3> CylindricalBinaryCompactObject::create_domain() const {
+  size_t first_inner_sphere_block = 46;
+  if (include_inner_sphere_A_) {
+    for (size_t i = 0; i < 10; ++i) {
+      abutting_directions_A.emplace(first_inner_sphere_block + i,
```
Possible to satisfy all the codecov complains here?

geoffrey4444 reviewed (changes requested):
Happy to approve, but I see codecov failures and wanted to make sure we don't care about those first. No other code changes to suggest

markscheel commented:
Rebased and added directives to silence codecov in a separate commit.

nilsvu commented on tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp:
```diff
@@ -283,9 +284,63 @@ void test_connectivity_once(const bool with_sphere_e,
+    if(include_inner_sphere_B) {
+      const size_t first_inner_sphere_block = include_inner_sphere_A ? 60 : 46;
+      for (size_t i = first_inner_sphere_block;
+           i < first_inner_sphere_block + 10; ++i) {
```
This test just reimplements the code that we're testing, doesn't it? That's not super useful and hard to maintain. Can we add some consistency checks to `test_domain_creator` instead, e.g. that the block has an external boundary in that direction etc?

markscheel replied:
I realize that this test (like many other tests) are re-implementations of the same code in a different way, but I'm not sure what you want me to do. This is basically the same excision_spheres test as in Test_BinaryCompactObject (but the case here is a bit more complicated).  `test_domain_creator` is general, and the property of which blocks have which excision boundaries in which directions are domain-dependent, e.g. they are different for BinaryCompactObject and CylindricalBinaryCompactObject, and they are different depending on the options.

nilsvu replied:
My suggestion is to replace this reimplementation with a consistency check in `test_domain_creator` that all `abutting_directions()` in all `DomainCreator::excision_spheres()` are also `Block::external_boundaries()`. That should be enough. Maybe there's a way to make sure that the excision sphere is fully covered with blocks, but I can't think of a nice way to do that right now.

nilsvu replied:
And you're right about the "like many other tests": reimplementations are done all over the place in the domain creator tests, which make them very hard to maintain. I opened an issue for that here: #4408

nilsvu commented on src/Domain/Creators/CylindricalBinaryCompactObject.cpp:
```diff
@@ -801,7 +810,70 @@ Domain<3> CylindricalBinaryCompactObject::create_domain() const {
+      // LCOV_EXCL_START
+      abutting_directions_B.emplace(first_inner_sphere_block + 10 + i,
+                                    Direction<3>::lower_xi());
+      // LCOV_EXCL_STOP
```
Unless this is an lcov bug (is it?), please don't add lcov excl statements to code that could reasonably be called in a test. The codecov CI test is not required to merge PRs so we can decide to ignore it.

markscheel replied:
I think this code *is* called in the test (unless I messed something up somewhere).  So I would consider this an LCOV bug.

nilsvu replied:
Yea this is weird. It worries me a bit because I haven't seen such an obvious lcov bug before. Well I have no change request here 🤷‍♂️

markscheel committed:
Add excision_spheres test to test_domain_creator

Tests that every direction of every excision sphere is
an external boundary of the correct Block.

markscheel committed:
Fix bug in Shell.

ExcisionSpheres were set independent of which_wedges_.

markscheel commented:
I added a new test to test_domain_creator.  This revealed a bug in Shell, which I fixed.  Removed the old test.

nilsvu commented on tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp:
```diff
@@ -287,60 +286,6 @@ void test_connectivity_once(const bool with_sphere_e,
-        ExcisionSphere<3>{inner_radius_objectA, center_objectA,
-                          abutting_directions_A});
-
-    CHECK(domain.excision_spheres() == expected_excision_spheres);
```
I'd say keep the check that the number of excision spheres is 2, and also their name, radius, and center.

nilsvu reviewed (commented):
LGTM please squash (including this)

markscheel commented:
Made change and squashed.  Left the Shell commit and the new-test commit separate since they are logically distinct.

nilsvu reviewed (dismissed):
Thanks for adding the consistency test 👍

markscheel commented:
I rebased onto develop because there was a conflict.
